Adding user debugging docs (#287)
PennyLaneAI/catalyst: doc/dev/sharp_bits.rst

optimization to take place within Catalyst:

.. code-block:: python

    @qjit
    def optimize(init_weights, data, steps):
        def loss(x):
            dy = grad(cost, argnum=0)(x, data)[0]
            return (cost(x, data), dy)

        opt = jaxopt.GradientDescent(loss, stepsize=0.4, value_and_grad=True)
        update_step = lambda i, *args: tuple(opt.update(*args))

        params = init_weights
        state = opt.init_state(params)

        return for_loop(0, steps, 1)(update_step)(params, state)[0]

The optimization now takes 574ms ± 43.1ms to complete when using 200 steps.
Note that, to compute gradients within a qjit-compiled function,
the :func:`catalyst.grad` function must be used.

JAX support and restrictions
----------------------------

Catalyst is utilizes JAX for program capture, which means you are able to
leverage the many functions accessible in ``jax`` and ``jax.numpy`` to write
code that supports :func:`@qjit <~.qjit>` and dynamic variables.

Currently, we are aiming to support as many JAX functions as possible, however
there may be cases where there is missing coverage. Known JAX functionality
that doesn't work with Catalyst includes:

> - ``jax.numpy.polyfit``

Author: ```suggestion
- ``jax.numpy.polyfit``
- ``jax.numpy.fft``
```

Debugging and sharp bits

> ========================

Reviewer: I was thinking a table of content might be nice to jump to the relevant sections and see at a glance what topics are covered.

Author: There is a table of contents automatically on the rhs!

Reviewer: I don't see it anywhere in the preview 👀

Reviewer: ![image](https://github.com/PennyLaneAI/catalyst/assets/29467667/e2f186ca-d530-49b8-9fd4-53062506a631)

Reviewer: Maybe this a screen size thing?

Author: Ah, yes - your window might be slightly too narrow 🤔

Author: This is a story I can make for the cloud team to fix

> Debugging and sharp bits
> ========================

Author: ```suggestion
Sharp bits and debugging tips
=============================
```

Catalyst is designed to allow you to take the tools and code patterns you are
familiar with when exploring quantum computing (such as Python, NumPy, JAX,
and PennyLane), while unlocking faster execution and the ability to run
hybrid quantum-classical workflows on accelerator devices.

Similar to JAX, Catalyst does this via the :func:`@qjit <.qjit>` decorator, which captures
hybrid programs written in Python, PennyLane, and JAX, and compiles them to

> native machine code --- preserving important aspects like conditional

Reviewer 2: It might be nice to be more precise with what you mean by native machine code. The reader might think we support some QPUs.

Author: Good point, do you have a suggestion? 🤔

> native machine code --- preserving important aspects like conditional
> branches and classical control.

Reviewer 2: ```suggestion
native machine code --- preserving control flow like conditional branches and loops.
```

> With Catalyst, we aim to try and support as many idiomatic PennyLane and JAX

Reviewer 2: ```suggestion
With Catalyst, we aim support as many idiomatic PennyLane and JAX
```

restrictions and constraints that should be taken into account**.

Here, we aim to provide an overview of the restrictions and constraints
(the 'sharp bits'), as well as debugging tips and common patterns that are
helpful when using Catalyst.

.. note::

    For a more general overview of Catalyst, please see the
    :doc:`quick start guide <quick_start>`.

Debugging functions

> -------------------

Author: Let's remove it for now.

> Debugging functions
> -------------------
> 
> Catalyst provides the following functions to help with debugging:
> 
> .. raw:: html
> 
>     <div class="summary-table">
> 
> .. autosummary::
>     :nosignatures:
> 
>     ~catalyst.debug.print
>     ~catalyst.autograph_source
> 
> .. raw:: html
> 
>     </div>
>     <div class="summary-table">

Author: ```suggestion
```

our :func:`@qjit <~.qjit>` compiled function:

>>> @qjit
... def f(x):
...     print(f"x = {x}")
...     return x ** 2
>>> f(2.)
x = Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
array(4.)
>>> f(3.)
array(9.)

We can see that on the first execution, program capture/tracing occurs, and we
can see the dynamic variable is printed (tracers capture *type*
and *shape*, but not numeric value). This captured program is compiled, and
then the binary is executed directly to return the function value --- the
print statement is never invoked with the numerical value of ``x``.

When we execute the function again, steps (1) and (2) are skipped since we
have already compiled a binary; this is called directly to get the function
result, and again the print statement is never hit.

This allows us to distinguish between computations that happen
at **compile-time** (steps 1 and 2), such as the ``print`` statement above,
and those that happen at **runtime** (step 3).

.. note::

    As a general rule of thumb, for a function that is repeatedly executed
    with different parameters, we want as much evaluation as possible to

>     happen at compile time.

Author: I think the JAX docs also say this somewhere :) @dime10 do you want any changes here?

>     happen at compile time.

Reviewer: Do you know where it says that?

One reason JAX introduced omnistaging is so that expensive operations on constants are still staged out to accelerators. So I'm not quite sure what to make of your statement.

>     happen at compile time.

Author: Oh good point. I can delete it 👍 

Before I do though, is there any (very general) rule of thumb you would like to have people keep in their mind while writing catalyst programs?

>     happen at compile time.

Reviewer: Hmm, in the context of this paragraph, maybe that generally things that happen at compile-time are slow (or lead to slowdowns), while things that happen at runtime are fast (or lead to speadups). (Think compile-time vs runtime loops, computing in Python vs in native code).

However, if the same computation is repeated every time the compiled function is run (where the results are the same no matter the inputs), and it is expensive, then it may be worth doing the computation once in Python and use the results statically in the program (for example Hamiltonian/chemistry calculations?).

>     As a general rule of thumb, for a function that is repeatedly executed
>     with different parameters, we want as much evaluation as possible to
>     happen at compile time.

Author: ```suggestion
    As a general rule of thumb, things that happen at compile-time
    are slow (or lead to slowdowns), while things that happen at
    runtime are fast (or lead to speadups).

    However, if the same computation is repeated every time the
    compiled function is run (where the results are the same no
    matter the inputs), and it is expensive, then it may be worth
    doing the computation once in Python and use the results
    statically in the program.
```

Perfect 💯

    happen at compile time.

    However, computations at compile time cannot depend on the value of
    dynamic variable, since this is not known yet. It can only depend
    on **static variables**, those whose values are known.

.. note::

    A general guideline when working with JIT compilation and Catalyst:

    - Python control flow and third party libraries like NumPy and SciPy will
      be evaluated at compile-time, and can only accept static variables.

    - JAX functions, such as ``jax.numpy``, and Catalyst functions like
      :func:`~.cond` and :func:`~.for_loop` will be evaluated at
      runtime, and can accept dynamic variables.

>     Note that if AutoGraph is enabled, Catalyst will attempt to convert Python

Reviewer 2: I would maybe move that to the first point

Author: Hmm, I tried but it makes it a bit harder to read that way 🤔

>     control flow to its Catalyst equivalent to support dynamic variables.

Reviewer 3: I first suggested adding the TF and Catalyst import ordering issue here in this doc page, but I think this should be added as a warning to the AutoGraph guide page as the AutoGraph support is still an experimental support and recalled in the doc in "note" blocks. What do you think @josh146 ?

Author: I agree :)

This concrete value was not available in Python because it depends on the value of the argument x.
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerBoolConversionError

This function will fail, as the Python ``if`` statement cannot accept a
dynamic variable (a JAX tracer) as an argument.

Instead, we can use Catalyst control flow :func:`~.cond` here:

>>> @qjit
... def f(x):
...     @cond(x > 5.)
...     def g():
...         return x / 2
...     @g.otherwise
...     def h():
...         return x
...     return g() ** 2
>>> f(2.)
array(4.)
>>> f(6.)
array(9.)

Here, both conditional branches are compiled, and only evaluated at runtime
when the value of ``x`` is known.

.. note::

    AutoGraph is an experimental feature that converts Python control flow
    that depends on dynamic variables to Catalyst control flow behind the
    scenes:

    >>> @qjit(autograph=True)
    ... def f(x):
    ...     if x > 5.:
    ...         print(x)
    ...         x = x / 2
    ...     return x ** 2
    >>> Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
    ... array(4.)
    >>> f(6.)
    ... array(9.)

>     For more details, see the AutoGraph guide.

Author: I might leave in just for now, as I'm going to try and write the guideline tomorrow 😬

Note that, if the Python ``if`` statement depends only on values that are
static (known at compile time), this is fine --- the ``if`` statement will
simply be evaluated at compile time rather than runtime:

>>> @qjit
... def f(x):
...     for i in range(2):
...         print(i, x)
...         x = x / 2
...     return x ** 2
>>> f(2.)
0 Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
1 Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
array(0.25)

Here, the for loop is evaluated at compile time (notice the multiple tracers
that have been printed out during program capture --- one for each loop!),

> rather than runtime.

Author: @dime10 yes I wrote this explicitly ignoring Autograph - what changes do you suggest here?

> rather than runtime.

Author: For me, I think the main valuable thing was knowing that `for ...:` etc. _may_ still work without autograph, but the behaviour may not be what you intend

> rather than runtime.

Reviewer: Oh I see, I think because the preceding note was talking about Autograph I thought this was still in the same context.

Maybe we could move the autograph note to after this paragraph?

Consider the following PennyLane example, where we have a parametrized
circuit, are measuring an expectation value, and are optimizing the result:

.. code-block:: python

    dev = qml.device("default.qubit", wires=4)

    @qml.qnode(dev)
    def cost(weights, data):
        qml.AngleEmbedding(data, wires=range(4))

        for x in weights:
            # each trainable layer
            for i in range(4):
                # for each wire
                if x[i] > 0:
                    qml.RX(x[i], wires=i)
                elif x[i] < 0:
                    qml.RY(x[i], wires=i)

            for i in range(4):
                qml.CNOT(wires=[i, (i + 1) % 4])

        return qml.expval(qml.PauliZ(0) + qml.PauliZ(3))

    weights = jnp.array(2 * np.random.random([5, 4]) - 1)
    data = jnp.array(np.random.random([4]))

    opt = jaxopt.GradientDescent(cost, stepsize=0.4)

    params = weights
    state = opt.init_state(params)

    for i in range(200):
        (params, _) = tuple(opt.update(params, state, data))

Using PennyLane v0.32 on Google Colab with the Python 3 Google Compute Engine
backend, this optimization takes 3min 28s ± 2.05s to complete.

> Let's switch over to Lightning, our high-performance statevector simulator,

Reviewer 3: Would you like to link to the Lightning repo as you're adding a brief introduction here or this is too much? :D

Author: ```suggestion
Let's switch over to [Lightning](https://docs.pennylane.ai/projects/lightning/en/stable/), our high-performance statevector simulator,
```

alongside the adjoint differentiation method. To do so, we change the first
two lines of the above code-block to set the device as ``"lightning.qubit"``,
and specify ``diff_method="adjoint"`` in the QNode decorator. With this
change, we have reduced the execution time down to 30.7s ± 1.8s.

We can rewrite this QNode to use Catalyst control flow, and compile
it using Catalyst:

.. code-block:: python

    dev = qml.device("lightning.qubit", wires=4)

    @qjit
    @qml.qnode(dev)
    def cost(weights, data):
        qml.AngleEmbedding(data, wires=range(4))

        def layer_loop(i):
            x = weights[i]
            def wire_loop(j):

                @cond(x[j] > 0)
                def trainable_gate():
                    qml.RX(x[j], wires=j)

                @trainable_gate.else_if(x[j] < 0)
                def negative_gate():
                    qml.RY(x[j], wires=j)

                trainable_gate.otherwise(lambda: None)
                trainable_gate()

            def cnot_loop(j):
                qml.CNOT(wires=[j, jnp.mod((j + 1), 4)])

>             for_loop(0, 4, 1)(wire_loop)()

Reviewer 3: I think it's good that unnecessary library imports are removed. But, `cond`, `for_loop`, and `grad` in these examples may be a bit confusing in particular for users new to the control-flow support in Catalyst. What about adding `from catalyst import ...` to these examples?

Author: Hmm I might leave this for now, since it matches the convention in our docstrings. But good point, we should revisit this.

        return qml.expval(qml.PauliZ(0) + qml.PauliZ(3))

    opt = jaxopt.GradientDescent(cost, stepsize=0.4)

    params = weights
    state = opt.init_state(params)

    for i in range(200):
        (params, _) = tuple(opt.update(params, state, data))

With the quantum function qjit-compiled, the optimization loop
now takes 16.4s ± 1.51s.

However, while the quantum function is now compiled, and the compiled function
is called to compute cost and gradient values, the optimization loop is still
occuring in Python.

Instead, we can write the optimization loop itself as a function and decorate
it with ``@qjit``; this will compile the optimization loop, and allow the full
optimization to take place within Catalyst:

.. code-block:: python

    @qjit
    def optimize(init_weights, data, steps):
        def loss(x):
            dy = grad(cost, argnum=0)(x, data)[0]
            return (cost(x, data), dy)

        opt = jaxopt.GradientDescent(loss, stepsize=0.4, value_and_grad=True)
        update_step = lambda i, *args: tuple(opt.update(*args))

        params = init_weights
        state = opt.init_state(params)

        return for_loop(0, steps, 1)(update_step)(params, state)[0]

> The optimization now takes 574ms ± 43.1ms to complete when using 200 steps.

Reviewer 2: Very nice example :+1:

> the :func:`catalyst.grad` function must be used.

Reviewer: Ah, I see this is covered in the section further down 🤔

> Note that, to compute gradients within a qjit-compiled function,
> the :func:`catalyst.grad` function must be used.

Author: haha I think I only discovered `jax.grad` worked _after_ I wrote this, hence the contradiction lol
```suggestion
Note that, to compute hybrid quantum-classical gradients within a qjit-compiled function,
the :func:`catalyst.grad` function must be used.
```

  indices.

If you come across any other JAX functions that don't work with Catalyst
(or don't already have a Catalyst equivalents), please let us know by opening
a `GitHub issue <https://github.com/PennyLaneAI/catalyst/issues>`__.

While leveraging ``jax.numpy`` makes it easy to port over NumPy-based
PennyLane workflows to Catalyst, we also inherit `various restrictions
and 'gotchas' from JAX
<https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html>`__.
This includes:

* **Pure functions**: Compilation is primarily designed to only work on pure
  functions. That is, functions that do not have any side-effects; the
  output is purely dependent only on function inputs.

* **In-place array updates**: Rather than using in-place array updates, the
  syntax ``new_array = jax_array.at[index].set(value)`` should be used. For
  more details, see `jax.numpy.ndarray.at
  <https://jax.readthedocs.io/en/latest/_autosummary /jax.numpy.ndarray.at.html>`__.

* **Lack of stateful random number generators**: In JAX, random number
  generators need to be explicitly created within the :func:`@qjit <~.qjit>` function

>   using ``jax.random.PRNGKey(int)``:

Reviewer: I don't think the `within the qjit function` part is accurate, the key can also both, be created as a constant outside the function (closure), and passed in as an argument (a JAX prng key is just a regular array).

I think what might also be important to show here is the "stateless" nature of the PRNG, so the key state must be explicitly updated each time you want to compute a random number:
```py
key = jax.random.PRNGKey(seed)

key, subkey = jax.random.split(key)
r1 = jax.random.normal(subkey, shape=(1,))

key, subkey = jax.random.split(key)
r2 = jax.random.normal(subkey, shape=(1,))
```

It would also be good to link to the JAX docs here, maybe
https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html
or
https://jax.readthedocs.io/en/latest/jax.random.html

Author: ```suggestion
* **Lack of stateful random number generators**: In JAX, random number
  generators are stateless, and the key state must be explicitly updated each time you want to compute a random number. For more details, see the `JAX documentation <https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html>`__.
```

  >>> @qjit()
  ... def f():
  ...     key = jax.random.PRNGKey(0)
  ...     a = jax.random.normal(key, shape=(1,))
  ...     return a
  >>> f()
  array([-0.78476578])

* **Dynamic-shaped arrays:** Functions that create or return arrays with
  dynamic shape --- that is, arrays where their shape is determined by a
  dynamic variable at runtime -- are currently not supported in JAX nor
  Catalyst. Typically, workarounds involve rewriting the code to utilize

>   ``jnp.where`` where possible.

Reviewer: Is there a reference for this? I don't think I would know what to do with this hint.

Author: @dime10 I don't have a good example no :( This is a tip that they also suggest in the JAX docs (try re-writing using `jnp.where`, which is in fact a wrapper around `lax.select`).

I've also used this in the past (successfully), but I remember it being non-trivial

Reviewer: Hmm I see. In this case we can leave it as is I think 👍

for classical processing, mixing ``jax.jit`` and ``qjit`` will allow this.
However, if possible, try to always use ``qjit`` to compile your entire
workflow.

Internal QJIT transformations
-----------------------------

Inside of a qjit-compiled function, JAX transformations
(``jax.grad``, ``jax.jacobian``, ``jax.vmap``, etc.)
can be used **as long as they are not applied to quantum processing**.

>>> @qjit
... def f(x):
...     def g(y):
...         return -jnp.sin(y) ** 2
...     return jax.grad(g)(x)
>>> f(0.4)
array(-0.71735609)

If they are applied to quantum processing, an error will occur:

>>> @qjit
... def f(x):
...     @qml.qnode(dev)
...     def g(y):
...         qml.RX(y, wires=0)
...         return qml.expval(qml.PauliX(0))
...     return jax.grad(lambda y: g(y) ** 2)(x)
>>> f(0.4)
NotImplementedError: must override

Instead, only Catalyst transformations will work when applied to hybrid
quantum-classical processing:

>>> @qjit
... def f(x):
...     @qml.qnode(dev)
...     def g(y):
...         qml.RX(y, wires=0)
...         return qml.expval(qml.PauliZ(0))
...     return grad(lambda y: g(y) ** 2)(x)
>>> f(0.4)
array(-0.71735609)

> Always use the equivalent Catalyst transformation

Reviewer 2: Do we have a performance comparison?

Author: We don't, no! Interestingly, I imagine `jax.grad` might be slower, since the transformation happens in Python/jaxpr, while `catalyst.grad` the transformation happens in MLIR 🤔

Author: Compilation a bit slower:

![image](https://github.com/PennyLaneAI/catalyst/assets/2959003/58a447ea-0fc9-4d61-9288-52623738016a)

But resulting binary approximately no difference:

![image](https://github.com/PennyLaneAI/catalyst/assets/2959003/f48e53ef-57ec-49e3-b554-0edee951e931)

Reviewer: Very cool! The "differentiation process" in Python need not necessarily be slower since it happens at a very high level, Enzyme has a lot more instructions to process.

That the execution speed is slightly slower with Enzyme is interesting, it would be interesting to investigate how this shakes out on larger examples.
I remember asking people at the Enzyme conference how the performance compares to a framework like JAX, but the answers weren't so clear. Enzyme proposes that differentiation _can_ produce faster code if you perform it after code optimizations, but in practice it might be difficult to beat high-level AD.

Author: @dime10 yeah would be interesting to share this with the enzyme team :)

Author: >That the execution speed is slightly slower with Enzyme is interesting, it would be interesting to investigate how this shakes out on larger examples.

I think I tested some larger examples after I posted this comment, and it did in fact seem to scale worse. E.g., take the example above but execute with a 100000 sized array, and enzyme was ~2-5x slower iirc

(:func:`catalyst.grad`, :func:`catalyst.jacobian`, :func:`catalyst.vjp`, :func:`catalyst.jvp`)
inside of a qjit-compiled function.

Inspecting and drawing circuits
-------------------------------

A useful tool for debugging quantum algorithms is the ability to draw them. Currently,

> :func:`@qjit <~.qjit>` compiled QNodes used as input to

Reviewer: ```suggestion
:func:`@qjit <~.qjit>` compiled QNodes can be used as input to
```

:func:`qml.draw <pennylane.draw>`, with the following caveats:

- :func:`qml.draw <pennylane.draw>` call must occur outside the :func:`@qjit <.qjit>`

> - The :func:`@qjit <.qjit>` decorator must be placed directly on top of the QNode

Reviewer: I think this is slightly misleading, you should be able to draw QNodes even if they don't have a qjit decorator, as in:
```py
@qml.qnode(dev)
def circuit(x):
    qml.Hadamard(0)
    return qml.expval(qml.PauliZ(0))
    
@qjit
def workflow(x):
    return circuit(x)
    
qml.draw(circuit)(x)
```
has always worked. This form was the recent addition:
 ```py
```py
@qjit
@qml.qnode(dev)
def circuit(x):
    qml.Hadamard(0)
    return qml.expval(qml.PauliZ(0))

qml.draw(circuit)(x)
```

Author: ```suggestion
- If used, the :func:`@qjit <.qjit>` decorator must be placed directly on top of the QNode (that is, you cannot draw a hybrid `qjit` compiled function that _contains_ a QNode).
```

Reviewer: I'm still not quite sure what the intended message is. Do you have an example for the situation that you are trying to exclude?

Author: It's maybe a bit obvious, but I'm trying to exclude two things:

```python
@qml.qnode(dev)
def circuit(x):
    ...

@qjit
def function(x):
    qml.draw(circuit)(x)
```

and 

```python
@qjit
def hybrid_function(x):
    @qml.qnode(dev)
    def circuit(x):
        ...

qml.draw(hybrid_function)(x)
```

Both are super obviously not going to work, but both tripped me up in the past.

Author: so I'm basically trying to say 'you can only draw circuits of the form `qjit(qnode)` or `qnode`; don't try and draw anything else

Reviewer: Oh interesting, would the second one work in pennylane?

Author: Nope it doesn't work either 😆 Don't know why I kept trying to do it.

- The :func:`catalyst.measure` function is not supported in drawn QNodes

- Catalyst conditional functions, such as :func:`~.cond` and
  :func:`~.for_loop`, will be 'unrolled'. That is, the drawn circuit will
  be a straight-line circuit, without any of the control flow represented
  explicitly.

For example,

.. code-block:: python

    @qjit
    @qml.qnode(dev)
    def circuit(x):
        def measurement_loop(i, y):
            qml.RX(y, wires=0)
            qml.RY(y ** 2, wires=1)
            qml.CNOT(wires=[0, 1])

            @cond(y < 0.5)
            def cond_gate():
                qml.CRX(y * jnp.exp(- y ** 2), wires=[0, 1])

            cond_gate()

            return y * 2

        for_loop(0, 3, step=1)(measurement_loop)(x)
        return qml.expval(qml.PauliZ(0))

>>> print(qml.draw(circuit)(0.3))
0: ──RX(0.30)─╭●─╭●─────────RX(0.60)─╭●──RX(1.20)─╭●─┤  <Z>
1: ──RY(0.09)─╰X─╰RX(0.27)──RY(0.36)─╰X──RY(1.44)─╰X─┤

At the moment, additional PennyLane `circuit inspection functions
<https://docs.pennylane.ai/en/stable/introduction/inspecting_circuits.html>`__
are not supported with Catalyst.

> Classical control debugging

Reviewer: Similar to my comment further up I would suggest `conditional` when talking about if/else statements, and control flow in general when talking about if/else/while/for...

> Classical control debugging
> ---------------------------

Author: ```suggestion
Conditional debugging
---------------------
```

.. note::

>     See our AutoGraph guide for converting native Python control flow

Reviewer 3: A link to the guide?

>     See our AutoGraph guide for converting native Python control flow
>     to QJIT compatible control.

Reviewer: I want to avoid the notion that this might be manual process 😅 
```suggestion
    See our AutoGraph guide for seamless conversion of native Python control flow
    to QJIT compatible control flow.
```

  >>> @qjit
  ... def f(x: float):
  ...     @cond(x > 1.5)
  ...     def cond_fn():
  ...         return x ** 2  # float
  ...     @cond_fn.otherwise
  ...     def else_branch():
  ...         return 6.  # float
  ...     return cond_fn()
  >>> f(1.5)
  array(6.)

- Similarly, the else (``my_cond_fn.otherwise``) may be omitted **as long as
  other branches do not return any values**. If other branches do return values,
  the else branch must be specified.

  >>> @qjit
  ... def f(x: float):
  ...     @cond(x > 1.5)
  ...     def cond_fn():
  ...         return x ** 2
  ...     return cond_fn()
  TypeError: Conditional requires consistent return types across all branches, got:
  - Branch at index 0: [ShapedArray(float64[], weak_type=True)]
  - Branch at index 1: []
  Please specify an else branch if none was specified.

  >>> @qjit
  ... def f(x: float):
  ...     @cond(x > 1.5)
  ...     def cond_fn():
  ...         return x ** 2
  ...     @cond_fn.otherwise
  ...     def else_branch():
  ...         return x
  ...     return cond_fn()
  >>> f(1.6)
  array(2.56)

- Finally, a reminder that conditional functions provided to :func:`~.cond` cannot
  accept any arguments.

> 

Reviewer: Should there be a section about loops then if we have one about conditionals?

Author: Maybe, but I don't have any tips for for loops 😅 something we can add after?

Reviewer: Maybe the thing about the iteration variable which is automatically inserted for for loops? :)

Compatibility with PennyLane transforms
---------------------------------------

PennyLane provides a wide variety of
:doc:`transforms <code/qml_transforms>` that
convert a circuit to one or more circuits.

> As a general rule of thumb, transforms that result in a single circuit are
> generally applied before the QNode decorator, while transforms that result in
> multiple circuits to be executed (**batch transforms**, such as
> :func:`~pennylane.transforms.split_non_commuting`) are applied after the QNode decorator.

Author: ```suggestion
As a general rule of thumb, transforms that result in a single circuit are
generally applied before/outside the QNode decorator, while transforms that result in
multiple circuits to be executed (**batch transforms**, such as
:func:`~pennylane.transforms.split_non_commuting`) are applied after/inside the QNode decorator.
```

> Currently, batch transforms and transforms that apply *after* the QNode decorator will
> not work with Catalyst:

Author: ```suggestion
Currently, batch transforms and transforms that apply *inside* or *within* the QNode decorator will
not work with Catalyst:
```

.. code-block:: python

    @qjit
    @qml.transforms.split_non_commuting

>     @qml.qnode(dev)

Reviewer: I found the terms *before* and *after* the QNode decorator a bit confusing, because in the code they occur in opposite order than during execution.

Maybe *inside*/*outside* the qnode would be clearer, or some other terms? 🤔

    def circuit(x):
        qml.RX(x,wires=0)
        return [qml.expval(qml.PauliY(0)), qml.expval(qml.PauliZ(0))]

>>> circuit(0.4)
CompileError: QuantumCompilationPass failed.

However, transforms that are applied *before* the QNode decorator will work with
Catalyst, as long as:

- The circuit does not include any Catalyst-specific features, such
  as Catalyst control flow or measurement,

- AutoGraph is disabled, and

- The transformation does not require or depend on the numeric value of
  dynamic variables.

For example:

.. code-block:: python

    @qjit
    @qml.qnode(dev)

>     @qml.transforms.merge_rotations

Author: ```suggestion
    @qml.transforms.merge_rotations()
```

          e:f64[] = integer_pow[y=2] c
          f:f64[1] = broadcast_in_dim[broadcast_dimensions=() shape=(1,)] c
          g:f64[1] = broadcast_in_dim[broadcast_dimensions=() shape=(1,)] e
          h:f64[1] = add f g
          i:f64[1] = slice[limit_indices=(1,) start_indices=(0,) strides=(1,)] h
          j:f64[] = squeeze[dimensions=(0,)] i
          k:AbstractQbit() = qextract d 0
          l:AbstractQbit() = qinst[op=RX qubits_len=1] k j
          m:AbstractObs(num_qubits=None,primitive=None) = namedobs[kind=PauliZ] l
          n:f64[] = expval[shots=None] m
           = qdealloc d
        in (n,) }
      fn=<QNode: wires=1, device='lightning.qubit', interface='auto', diff_method='best'>
    ] a
  in (b,) }

Function argument restrictions
------------------------------

Compiled functions can accept arbitrary function arguments, as long as the
inputs can be represented as `Pytrees
<https://jax.readthedocs.io/en/latest/pytrees.html>`__ --- tree-like
structures built out of Python container objects such as lists, dictionaries,
and tuples --- where the *values* (leaf nodes) are compatible types.

Compatible types includes booleans, Python numeric types, JAX arrays,

> and PennyLane quantum operators.

Reviewer: Small suggestion that might improve clarity, but maybe we can lead with the compilation compatible types before mentioning PyTrees?

> and PennyLane quantum operators.

Reviewer: I'm not sure I would mention quantum operators here, they only work because they are PyTrees and are not themselves leaves.

Author: Hmm, I would err to include them. Although not technically accurate since they aren't leaves, good to let users know that they can be used

.. note::

>     Catalyst currently doesn't support string arguments to compiled functions
>     with the exception of dictionary keys.

Reviewer: I'd like to avoid the impression that this is a "current" restriction in Catalyst.
```suggestion
    As is the case with JAX, string values cannot be used as arguments to compiled functions
    in Catalyst, with the exception of their use as dictionary keys.
```

Reviewer: I'm also torn on the mention of strings as dictionary keys. Any PyTree class can accept strings as metadata, so using dictionaries is not really a way to pass strings to a compiled function 🤔

Author: ```suggestion
    Catalyst currently doesn't support string arguments to compiled functions.
```

... def circuit(x):
...     qml.RX(x, wires=0)
...     qml.RY(0.1, wires=0)
...     return qml.expval(qml.PauliZ(0))
>>> circuit(jnp.array([0.1, 0.2]))
Array([0.99003329, 0.97517033], dtype=float64)
>>> qjit(circuit)(jnp.array([0.1, 0.2]))
UnboundLocalError: local variable 'baseType' referenced before assignment

While not as flexible as true vectorized quantum operations, as a workaround
``jax.vmap`` can be used to allow for multi-dimensional **function**
arguments:

>>> jax.vmap(qjit(circuit))(jnp.array([0.1, 0.2]))
Array([0.99003329, 0.97517033], dtype=float64)

Note that ``jax.vmap`` cannot be used within a qjit-compiled function:

>>> qjit(jax.vmap(circuit))(jnp.array([0.1, 0.2]))
NotImplementedError: Batching rule for 'qinst' not implemented

Functionality differences from PennyLane
----------------------------------------

The ultimate aim with Catalyst will be the ability to prototype quantum algorithms
in Python with PennyLane, and easily scale up prototypes by simply adding ``@qjit``.
This will require that all PennyLane functionality behaves identically whether or not
the ``@qjit`` decorator is applied.

Currently, however, this is not the case in two places.

> - **Finite-shot measurement statistics**. The Catalyst-Lightning runtime does not

Reviewer 3: Hmm.. The Catalyst Runtime has support for finite-shot measurement statistics, but the Lightning simulators don't provide this support in their C++ interface and as the result this feature is not supported in Catalyst.  

```suggestion
- **Finite-shot measurement statistics**. The Catalyst Runtime does not
```

Author: @maliasadi I think I would prefer to keep `Lightning` here, since as you say, this is lightning that does not support it!

  ...     @cond(x > 1.5)
  ...     def cond_fn():
  ...         return x ** 2  # float
  ...     @cond_fn.otherwise
  ...     def else_branch():
  ...         return 6.  # float
  ...     return cond_fn()
  >>> f(1.5)
  array(6.)

- Similarly, the else (``my_cond_fn.otherwise``) may be omitted **as long as
  other branches do not return any values**. If other branches do return values,
  the else branch must be specified.

  >>> @qjit
  ... def f(x: float):
  ...     @cond(x > 1.5)
  ...     def cond_fn():
  ...         return x ** 2
  ...     return cond_fn()
  TypeError: Conditional requires consistent return types across all branches, got:
  - Branch at index 0: [ShapedArray(float64[], weak_type=True)]
  - Branch at index 1: []
  Please specify an else branch if none was specified.

  >>> @qjit
  ... def f(x: float):
  ...     @cond(x > 1.5)
  ...     def cond_fn():
  ...         return x ** 2
  ...     @cond_fn.otherwise
  ...     def else_branch():
  ...         return x
  ...     return cond_fn()
  >>> f(1.6)
  array(2.56)

- Finally, a reminder that conditional functions provided to :func:`~.cond` cannot
  accept any arguments.

Compatibility with PennyLane transforms

> ---------------------------------------

Author: @erick-xanadu this is based on the current `main` branch, so will need to be updated when your PR is merged

          f:f64[1] = broadcast_in_dim[broadcast_dimensions=() shape=(1,)] c
          g:f64[1] = broadcast_in_dim[broadcast_dimensions=() shape=(1,)] e
          h:f64[1] = add f g
          i:f64[1] = slice[limit_indices=(1,) start_indices=(0,) strides=(1,)] h
          j:f64[] = squeeze[dimensions=(0,)] i
          k:AbstractQbit() = qextract d 0
          l:AbstractQbit() = qinst[op=RX qubits_len=1] k j
          m:AbstractObs(num_qubits=None,primitive=None) = namedobs[kind=PauliZ] l
          n:f64[] = expval[shots=None] m
           = qdealloc d
        in (n,) }
      fn=<QNode: wires=1, device='lightning.qubit', interface='auto', diff_method='best'>
    ] a
  in (b,) }

Function argument restrictions
------------------------------

Compiled functions can accept arbitrary function arguments, as long as the
inputs can be represented as `Pytrees
<https://jax.readthedocs.io/en/latest/pytrees.html>`__ --- tree-like
structures built out of Python container objects such as lists, dictionaries,
and tuples --- where the *values* (leaf nodes) are compatible types.

Compatible types includes Booleans, Python numeric types, JAX arrays,
and PennyLane quantum operators.

.. note::

>     Catalyst currently doesn't support string arguments to compiled functions.

Author: ```suggestion
    Non-numeric types, such as strings, are generally not supported as arguments to compiled functions.
```

      :func:`~.cond` and :func:`~.for_loop` will be evaluated at
      runtime, and can accept dynamic variables.

    Note that if AutoGraph is enabled, Catalyst will attempt to convert Python
    control flow to its Catalyst equivalent to support dynamic variables.

For example, consider the following:

>>> @qjit
... def f(x):
...     if x > 5:
...       x = x / 2
...     return x ** 2
>>> f(2.)
TracerBoolConversionError: Attempted boolean conversion of traced array with shape bool[]..
The error occurred while tracing the function f at <ipython-input-15-2aa7bf60efbb>:1 for make_jaxpr.
This concrete value was not available in Python because it depends on the value of the argument x.
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerBoolConversionError

This function will fail, as the Python ``if`` statement cannot accept a
dynamic variable (a JAX tracer) as an argument.

Instead, we can use Catalyst control flow :func:`~.cond` here:

>>> @qjit
... def f(x):
...     @cond(x > 5.)
...     def g():
...         return x / 2
...     @g.otherwise
...     def h():
...         return x
...     return g() ** 2
>>> f(2.)
array(4.)
>>> f(6.)
array(9.)

> Here, both conditional branches are compiled, and only evaluated at runtime
> when the value of ``x`` is known.

Author: ```suggestion
Here, both conditional branches are compiled, and only evaluated at runtime
when the value of ``x`` is known.

Note that, if the Python ``if`` statement depends only on values that are
static (known at compile time), this is fine --- the ``if`` statement will
simply be evaluated at compile time rather than runtime:

Let's consider an example where a for loop is evaluated at compile time:

>>> @qjit
... def f(x):
...     for i in range(2):
...         print(i, x)
...         x = x / 2
...     return x ** 2
>>> f(2.)
0 Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
1 Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
array(0.25)

Here, the for loop is evaluated at compile time (notice the multiple tracers
that have been printed out during program capture --- one for each loop!),
rather than runtime.
```

.. note::

    AutoGraph is an experimental feature that converts Python control flow
    that depends on dynamic variables to Catalyst control flow behind the
    scenes:

    >>> @qjit(autograph=True)
    ... def f(x):
    ...     if x > 5.:
    ...         print(x)
    ...         x = x / 2
    ...     return x ** 2
    >>> Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
    ... array(4.)
    >>> f(6.)
    ... array(9.)

    For more details, see the AutoGraph guide.

> Note that, if the Python ``if`` statement depends only on values that are
> static (known at compile time), this is fine --- the ``if`` statement will
> simply be evaluated at compile time rather than runtime:
> 
> >>> @qjit
> ... def f(x):
> ...     for i in range(2):
> ...         print(i, x)
> ...         x = x / 2
> ...     return x ** 2
> >>> f(2.)
> 0 Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
> 1 Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
> array(0.25)
> 
> Here, the for loop is evaluated at compile time (notice the multiple tracers
> that have been printed out during program capture --- one for each loop!),
> rather than runtime.

Author: ```suggestion
```

array(-0.71735609)

If they are applied to quantum processing, an error will occur:

>>> @qjit
... def f(x):
...     @qml.qnode(dev)
...     def g(y):
...         qml.RX(y, wires=0)
...         return qml.expval(qml.PauliX(0))
...     return jax.grad(lambda y: g(y) ** 2)(x)
>>> f(0.4)
NotImplementedError: must override

Instead, only Catalyst transformations will work when applied to hybrid
quantum-classical processing:

>>> @qjit
... def f(x):
...     @qml.qnode(dev)
...     def g(y):
...         qml.RX(y, wires=0)
...         return qml.expval(qml.PauliZ(0))
...     return grad(lambda y: g(y) ** 2)(x)
>>> f(0.4)
array(-0.71735609)

Always use the equivalent Catalyst transformation
(:func:`catalyst.grad`, :func:`catalyst.jacobian`, :func:`catalyst.vjp`, :func:`catalyst.jvp`)
inside of a qjit-compiled function.

Inspecting and drawing circuits
-------------------------------

A useful tool for debugging quantum algorithms is the ability to draw them. Currently,
:func:`@qjit <~.qjit>` compiled QNodes can be used as input to
:func:`qml.draw <pennylane.draw>`, with the following caveats:

- :func:`qml.draw <pennylane.draw>` call must occur outside the :func:`@qjit <.qjit>`

> - If used, the :func:`@qjit <.qjit>` decorator must be placed directly on top of the QNode (that is, you cannot draw a hybrid `qjit` compiled function that _contains_ a QNode).

Author: ```suggestion
- The `qml.draw()` function will only accept plain QNodes as input, _or_ QNodes that have been qjit-compiled. It will not accept arbitrary hybrid functions (that may contain QNodes).
```

Author: @dime10 rewrote this to make it clearer

Reviewer: Awesome, this one is very clear 💯

  indices.

If you come across any other JAX functions that don't work with Catalyst
(and don't already have a Catalyst equivalent), please let us know by opening
a `GitHub issue <https://github.com/PennyLaneAI/catalyst/issues>`__.

While leveraging ``jax.numpy`` makes it easy to port over NumPy-based
PennyLane workflows to Catalyst, we also inherit `various restrictions
and 'gotchas' from JAX
<https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html>`__.
This includes:

* **Pure functions**: Compilation is primarily designed to only work on pure
  functions. That is, functions that do not have any side-effects; the
  output is purely dependent only on function inputs.

* **In-place array updates**: Rather than using in-place array updates, the
  syntax ``new_array = jax_array.at[index].set(value)`` should be used. For
  more details, see `jax.numpy.ndarray.at
  <https://jax.readthedocs.io/en/latest/_autosummary /jax.numpy.ndarray.at.html>`__.

* **Lack of stateful random number generators**: In JAX, random number
  generators need to be explicitly created within the :func:`@qjit <~.qjit>` function
  using ``jax.random.PRNGKey(int)``:

> 
>   >>> @qjit()
>   ... def f():
>   ...     key = jax.random.PRNGKey(0)
>   ...     a = jax.random.normal(key, shape=(1,))
>   ...     return a
>   >>> f()
>   array([-0.78476578])

Author: ```suggestion
```
